cibuildwheel/__main__.py

```diff
@@ -144,11 +147,26 @@ def main() -> None:
         print(f"cibuildwheel: Unsupported platform: {platform}", file=sys.stderr)
         sys.exit(2)
 
-    options = compute_options(platform=platform, command_line_arguments=args)
+    # check if we have an sdist tarball as "package_dir" and extract first
+    package_dir = args.package_dir and Path(args.package_dir)
+    from_sdist = package_dir and package_dir.is_file()
+    if from_sdist:
+        if not str(package_dir).endswith(".tar.gz"):
+            msg = f"cibuildwheel: Package {package_dir} is not an sdist '.tar.gz' tarball"
+            print(msg, file=sys.stderr)
+            sys.exit(2)
+
+        extract_dir, _, _ = str(package_dir).rpartition(".tar.gz")
```

Comment:
We can't use [str.removesuffix()](https://docs.python.org/3/library/stdtypes.html#str.removesuffix) yet, but this still looks simpler to me:

```suggestion
        extract_dir = str(package_dir)[:-len(".tar.gz")]
```

Reply:
Actually the code here and below seems to make some assumptions that are not guaranteed, namely:
- that the tarball is located in the current directory (can be easily fixed by a `os.path.basename()`)
- that the tarball contains a directory with the same name as the archive (minus the `.tar.gz`)
- that the tarball doesn't have files that will be created outside of the c.w.d. (see the warning in https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractall )

I think we could at least verify that the tarball is "well-behaved" by calculating the common prefix directory of the members of the archive: if there is one, it can replace `extract_dir `here; otherwise, we just bail out.

Some inspiration can be taken e.g. from https://github.com/pypa/pip/blob/main/src/pip/_internal/utils/unpacking.py